### PR TITLE
Refactor common dialog/drawer code into a base class

### PR DESCRIPTION
**Pull request by m-akinc**
# Pull Request

## 🤨 Rationale

Refactor to avoid duplication between dialog and drawer.

## 👩‍💻 Implementation

Created common base class `DialogBase`.

## 🧪 Testing

Existing tests pass.

## ✅ Checklist

- [x] I have updated the project documentation to reflect my changes or determined no changes are needed.

### CODE REVIEW

**Comment by m-akinc** [comment]
@atmgrifter00 would you mind buddying this?

File: change/@ni-nimble-components-dc071140-b97c-473f-b929-1268a3aa3a40.json
Comment (atmgrifter00) [inline change/@ni-nimble-components-dc071140-b97c-473f-b929-1268a3aa3a40.json:2]: What makes this "minor" instead of "patch"?

**Review by atmgrifter00 — APPROVED** [review]

File: change/@ni-nimble-components-dc071140-b97c-473f-b929-1268a3aa3a40.json
Comment (m-akinc) [reply change/@ni-nimble-components-dc071140-b97c-473f-b929-1268a3aa3a40.json:2]: I wasn't sure, but I thought since we were technically exporting a new type (`DialogBase`) it should be "minor". But it's intended as an internal type, so I'm not sure.

**Comment by rajsite** [comment]
Adding @mollykreis to take a look since they are familiar with original code and the issue being addressed

File: packages/nimble-components/src/dialog-base/index.ts
Comment (m-akinc) [inline packages/nimble-components/src/dialog-base/index.ts:85]: This is the `CloseWatcher` mitigation. Basically everything else is just refactoring `Drawer` and `Dialog`.

File: packages/nimble-components/src/dialog/tests/dialog.spec.ts
Comment (mollykreis) [inline packages/nimble-components/src/dialog/tests/dialog.spec.ts:12]: Why do you need the `?? null` here? Isn't the type of `preventDismiss` a `boolean`, which would mean it is never not defined?

File: packages/nimble-components/src/drawer/index.ts
Comment (mollykreis) [inline packages/nimble-components/src/drawer/index.ts:103]: It doesn't look like `doneClosing` is called anywhere but here, so I'm curious why you decided to create a helper function to wrap two lines of this function.

**Review by mollykreis — APPROVED** [review]

File: packages/nimble-components/src/dialog/tests/dialog.spec.ts
Comment (m-akinc) [reply packages/nimble-components/src/dialog/tests/dialog.spec.ts:12]: You're right. I must have been confused.

File: packages/nimble-components/src/drawer/index.ts
Comment (m-akinc) [reply packages/nimble-components/src/drawer/index.ts:103]: It was for symmetry with the `startClosing` function that I added. One sets `closeReason` and sets `closing = true`, and the other uses `closeReason` and sets `closing = false`.

File: packages/nimble-components/src/dialog/styles.ts
Comment (m-akinc) [reply packages/nimble-components/src/dialog/styles.ts:50]: Good point. I'm now adding a 1px green border. I'm waiting on Brandon's approval.

Comment (m-akinc) [reply packages/nimble-components/src/dialog/styles.ts:49]: He wanted a 2px border instead, so I made that change.

**Review by jattasNI — COMMENTED** [review]
I didn't get a chance to review before going out of office. If other reviewers approve then you don't need to wait for me.

File: packages/nimble-components/src/drawer/index.ts
Comment (rajsite) [inline packages/nimble-components/src/drawer/index.ts:43]: I am just starting to look at it. My first impression is that if around half the methods of the base class are overridden without calling super for the default behavior then I'm not sure how much value we are getting in the abstraction / maybe the abstraction is misaligned. I'll have to review it more closely to understand the goals of the split / responsibilities of the base class.

File: packages/nimble-components/src/dialog-base/index.ts
Comment (rajsite) [inline packages/nimble-components/src/dialog-base/index.ts:19]: I'd probably frame this as "Modal" of which the Dialog and Drawer are implementations of. But like the other comment says haven't fully thought about what the specific responsibilities of the Modal base class should be

File: packages/nimble-components/src/drawer/index.ts
Comment (rajsite) [reply packages/nimble-components/src/drawer/index.ts:43]: @m-akinc to prevent blocking the functional aspect of the change, can we split the refactor and event handling updates into separate PRs?

Comment (m-akinc) [reply packages/nimble-components/src/drawer/index.ts:43]: Moved the functional aspects into PR #1848

Comment (jattasNI) [reply packages/nimble-components/src/drawer/index.ts:43]: I agree after scanning that the benefit is less compelling if only one of the derived classes uses many of the methods in the base class. Some general strategies around this:
1. don't put those methods in the base class. Each derived class has its own implementation. If, after doing this, the base class is mostly empty or isn't cohesive (e.g. has an `open` but not a `close`) then we could just get rid of it.
2. narrow the scope of those methods to only include the common functionality, then have the derived classes call super and then add their custom functionality
3. make the base class handle both cases and choose between them based on some configuration of the object (don't love this option, probably a mixin or larger class hierarchy would make more sense, but none of that feels right for this tiny class hierarchy)
4. find a way to make the behaviors identical so they can be handled entirely by the base class

@m-akinc wanna take a pass at applying one or more of these strategies?

Comment (rajsite) [reply packages/nimble-components/src/drawer/index.ts:43]: One thought I had is having Modal as the base class, the promise workflow is all handled there (we don't need implementation specific open and close value tracking and promise resolution). So things like resolveShow are private there and the dialog reference is protected (calling showModal on a dialog element is managed in the base class).

And then the base class has lifecycle methods that can be overridden / called and some state tracking. 

Something like states: {opening, open, closing, closed}
and methods (some protected) for the transitions: open, finishOpening, close, finishClosing.

So a dialog without animations in close will immediately call finishClosing while one with other work like animations can call finishClosing later. Think the logic will be clearer too tracking a state vs implied knowledge based on whether resolveShow is set or not.

Comment (m-akinc) [reply packages/nimble-components/src/drawer/index.ts]: > the dialog reference is protected

I guess it needs to be public for `ref` to work in the template.

**Review by jattasNI — APPROVED** [review]
@m-akinc FYI there are some lint errors but otherwise looks good!

File: packages/nimble-components/src/modal/index.ts
Comment (rajsite) [inline packages/nimble-components/src/modal/index.ts:15]: types should be a types.ts file

File: packages/nimble-components/src/drawer/index.ts
Comment (rajsite) [reply packages/nimble-components/src/drawer/index.ts]: Looking a lot better. The thing I want to emphasize was the goal "So things like resolveShow are private there" and "And then the base class has lifecycle methods that can be overridden / called and some state tracking."

The intention is that the ModalState tracking is completely handled in the base class and private. All modals go through the state transitions (off the top, didn't double check what was implemented) closed -> opening -> open -> closing -> closed. And the transitions between states are handled by methods that do the transitions finishOpening, finishClosing, etc.

Previously `resolveShow` represented the state of the control and now `state` represents the state of control. Either way the goal of my statement is that the state management is private / encapsulated in Modal.

File: packages/nimble-components/src/modal/index.ts
Comment (rajsite) [inline packages/nimble-components/src/modal/index.ts:75]: nit/pref: instead of accessing the public accessor inside the methods, which could have different meanings over time, lets compare the states instead. A nice syntax shortcut is something like `[ModalState.open, ModalState.closing].includes(this.state)`

File: packages/nimble-components/src/modal/index.ts
Comment (rajsite) [inline packages/nimble-components/src/modal/index.ts:51]: Should the dialog be considered "open" if the state is opening or closing? Do we have test coverage here?

File: packages/nimble-components/src/modal/index.ts
Comment (m-akinc) [reply packages/nimble-components/src/modal/index.ts:75]: Actually, we don't even have to check if closing. The open check is sufficient.

Comment (rajsite) [reply packages/nimble-components/src/modal/index.ts:75]: The type pattern we use for enums is to treat ModalState as a string. That's how we get type checked attribute strings in the angular templates, so the type `("open" | "closing")[]` is expected.

Looks like the surprising bit is the type for `includes` where I'd be on team the TypeScript is being too strict: https://github.com/Microsoft/TypeScript/issues/26255

So shortcut not worth the effort, just some good ol fashion `this.state === foo || this.state === bar`

File: packages/nimble-components/src/modal/index.ts
Comment (m-akinc) [reply packages/nimble-components/src/modal/index.ts:51]: It turns out that we do have a test that expects `open` to be true while closing. So I've changed `open` to mean `state !== closed`. Neither Dialog nor Drawer distinguish between `open` and `opening` states, though, so I don't think we need to test that the property `open` is true when `state === opening`.

File: packages/nimble-components/src/drawer/index.ts
Comment (m-akinc) [reply packages/nimble-components/src/drawer/index.ts]: > The intention is that the ModalState tracking is completely handled in the base class and private.

I had tried to do that, but it didn't seem like it would work. On second look, we can make it work with a couple assumptions:
1. Derived classes can override the `startOpening/Closing` methods, but should not call them (except from inside an override). I'm having `Modal` set the state immediately before calling those methods, so the state won't be set correctly if the derived class calls `startOpening` somewhere else. 
2. The base implementation for `finishOpening/Closing` sets the state, so if a derived class overrides one of those functions, it _must_ call the base implementation (or the state won't get updated properly).

I wish there were a nice way to enforce these assumptions, but instead I've just added comments to the base implementations.

File: packages/nimble-components/src/modal/index.ts
Comment (rajsite) [inline packages/nimble-components/src/modal/index.ts:132]: I think the issues here for needing a buunch of overrides are that the Modal class is trying to implement a specific workflow, where the transition is preset so i.e. startClosing immediately calls finishClosing. Instead there should be lifecycle hooks that each subclass is implementing.

startOpening is abstract and in the dialog implementation iit immediately calls finishOpeniing while in the drawer implementation it does its animations beforee calling finishClosing.

Comment (m-akinc) [reply packages/nimble-components/src/modal/index.ts]: I made `startOpening/Closing` abstract. It doesn't really change much, other than moving the default implementations to `Dialog`. I guess it technically reduces the number of overrides by 2 (if you don't count the abstract method implementations). That's all you wanted me to change, right?

File: packages/nimble-components/src/modal/index.ts
Comment (rajsite) [inline packages/nimble-components/src/modal/index.ts:85]: The responsibilities of cancel and close handlers don't seem quite right. I think similar to some previous commits they seem over coupled to specific workflows. Modal instances like drawer calling "super.cancelHandler" to start the transition to Closing and check the preventDismiss state seems like unclear responsibilities for a method called "cancelHandler".

Looking at it again I also have a more general question of why are we not designing animations into the Modal base class. I'm actually wondering now if it is really intentional that the dialog has no animations. Seems like even short, fast animations would be useful for all modals.

Some directions:
- We can continue to discuss and iterate on the responsibilities of the base class, including seeing if animation responsibilities make more sense there, it probably makes sense to discuss it first to prevent review cycles or continue the approach of aync comment and feedback if you want to keep going that way
- We pause the refactor for now since we don't have a clear need for shared Modal base classes at the moment (I'm not aware of new modals we need to implement or if we want to share more behavior like animations across all the modals)

Comment (m-akinc) [reply packages/nimble-components/src/modal/index.ts:85]: I'll de-prioritize and move to draft until I have new changes to review.

**Comment by rajsite** [comment]
Closing the draft as unprioritized right now. Think it would be useful to reconsider if adding additional modal types / decide to share more functionality (i.e. animations for dialogs and drawers)

**Pull request closed without merging**
